# Income Classification using ML and Python

## Data Importing

In [19]:
import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
%matplotlib inline


In [20]:
start=pd.read_csv('Income Classification.csv')

In [21]:
income_ds=pd.DataFrame(start)

In [22]:
## replacing ? with NaN value
income_ds.replace(to_replace=' ?',value=np.nan,inplace=True)

In [23]:
##displaying first 30 rows of dataset
income_ds.head(30)

age          workclass   fnlwgt      education   education-num  \
0    39          State-gov    77516      Bachelors              13   
1    50   Self-emp-not-inc    83311      Bachelors              13   
2    38            Private   215646        HS-grad               9   
3    53            Private   234721           11th               7   
4    28            Private   338409      Bachelors              13   
5    37            Private   284582        Masters              14   
6    49            Private   160187            9th               5   
7    52   Self-emp-not-inc   209642        HS-grad               9   
8    31            Private    45781        Masters              14   
9    42            Private   159449      Bachelors              13   
10   37            Private   280464   Some-college              10   
11   30          State-gov   141297      Bachelors              13   
12   23            Private   122272      Bachelors              13   
13   32            Private   205019     Assoc-acdm              12   
14   40            Private   121772      Assoc-voc              11   
15   34            Private   245487        7th-8th               4   
16   25   Self-emp-not-inc   176756        HS-grad               9   
17   32            Private   186824        HS-grad               9   
18   38            Private    28887           11th               7   
19   43   Self-emp-not-inc   292175        Masters              14   
20   40            Private   193524      Doctorate              16   
21   54            Private   302146        HS-grad               9   
22   35        Federal-gov    76845            9th               5   
23   43            Private   117037           11th               7   
24   59            Private   109015        HS-grad               9   
25   56          Local-gov   216851      Bachelors              13   
26   19            Private   168294        HS-grad               9   
27   54                NaN   180211   Some-college              10   
28   39            Private   367260        HS-grad               9   
29   49            Private   193366        HS-grad               9   

            marital-status          occupation    relationship  \
0            Never-married        Adm-clerical   Not-in-family   
1       Married-civ-spouse     Exec-managerial         Husband   
2                 Divorced   Handlers-cleaners   Not-in-family   
3       Married-civ-spouse   Handlers-cleaners         Husband   
4       Married-civ-spouse      Prof-specialty            Wife   
5       Married-civ-spouse     Exec-managerial            Wife   
6    Married-spouse-absent       Other-service   Not-in-family   
7       Married-civ-spouse     Exec-managerial         Husband   
8            Never-married      Prof-specialty   Not-in-family   
9       Married-civ-spouse     Exec-managerial         Husband   
10      Married-civ-spouse     Exec-managerial         Husband   
11      Married-civ-spouse      Prof-specialty         Husband   
12           Never-married        Adm-clerical       Own-child   
13           Never-married               Sales   Not-in-family   
14      Married-civ-spouse        Craft-repair         Husband   
15      Married-civ-spouse    Transport-moving         Husband   
16           Never-married     Farming-fishing       Own-child   
17           Never-married   Machine-op-inspct       Unmarried   
18      Married-civ-spouse               Sales         Husband   
19                Divorced     Exec-managerial       Unmarried   
20      Married-civ-spouse      Prof-specialty         Husband   
21               Separated       Other-service       Unmarried   
22      Married-civ-spouse     Farming-fishing         Husband   
23      Married-civ-spouse    Transport-moving         Husband   
24                Divorced        Tech-support       Unmarried   
25      Married-civ-spouse        Tech-support         Husband   
26           Never-married        Craft-repair     

In [24]:
#finding the shape i.e. no. of rows and columns of dataset
income_ds.shape

(32561, 15)

* There are 32561 rows and 15 attributes in the dataset.

In [25]:
#finding columns of the dataset
income_ds.columns

Index(['age', ' workclass', ' fnlwgt', ' education', ' education-num',
       ' marital-status', ' occupation', ' relationship', ' race', ' sex',
       ' capital-gain', ' capital-loss', ' hours-per-week', ' native-country',
       ' income'],
      dtype='object')

In [26]:
#info about the dataset
income_ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   age              32561 non-null  int64 
 1    workclass       30725 non-null  object
 2    fnlwgt          32561 non-null  int64 
 3    education       32561 non-null  object
 4    education-num   32561 non-null  int64 
 5    marital-status  32561 non-null  object
 6    occupation      30718 non-null  object
 7    relationship    32561 non-null  object
 8    race            32561 non-null  object
 9    sex             32561 non-null  object
 10   capital-gain    32561 non-null  int64 
 11   capital-loss    32561 non-null  int64 
 12   hours-per-week  32561 non-null  int64 
 13   native-country  31978 non-null  object
 14   income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 2.6+ MB


In [27]:
#descriptive analysis of data
income_ds.describe()

age        fnlwgt   education-num   capital-gain  \
count  32561.000000  3.256100e+04    32561.000000   32561.000000   
mean      38.581647  1.897784e+05       10.080679    1077.648844   
std       13.640433  1.055500e+05        2.572720    7385.292085   
min       17.000000  1.228500e+04        1.000000       0.000000   
25%       28.000000  1.178270e+05        9.000000       0.000000   
50%       37.000000  1.783560e+05       10.000000       0.000000   
75%       48.000000  2.370510e+05       12.000000       0.000000   
max       90.000000  1.484705e+06       16.000000   99999.000000   

        capital-loss   hours-per-week  
count   32561.000000     32561.000000  
mean       87.303830        40.437456  
std       402.960219        12.347429  
min         0.000000         1.000000  
25%         0.000000        40.000000  
50%         0.000000        40.000000  
75%         0.000000        45.000000  
max      4356.000000        99.000000

In [28]:
income_ds.isnull().sum()

age                   0
 workclass         1836
 fnlwgt               0
 education            0
 education-num        0
 marital-status       0
 occupation        1843
 relationship         0
 race                 0
 sex                  0
 capital-gain         0
 capital-loss         0
 hours-per-week       0
 native-country     583
 income               0
dtype: int64

* There are 3 attributes having Null values : workclass, occupation and native-country.

* workclass has 1836 null values,occupation has 1843 null values and native-country has 583 null values.

In [29]:
#Size of the Dataset
income_ds.size

488415

* There are 488415 data items in the Dataset.

In [30]:
#Datatypes of each attribute in the dataset
income_ds.dtypes

age                 int64
 workclass         object
 fnlwgt             int64
 education         object
 education-num      int64
 marital-status    object
 occupation        object
 relationship      object
 race              object
 sex               object
 capital-gain       int64
 capital-loss       int64
 hours-per-week     int64
 native-country    object
 income            object
dtype: object

* There are 6 attributes with int64 datatype and 9 attributes with object datatype.

In [31]:
#total non null values of each attribute
income_ds.count()

age                32561
 workclass         30725
 fnlwgt            32561
 education         32561
 education-num     32561
 marital-status    32561
 occupation        30718
 relationship      32561
 race              32561
 sex               32561
 capital-gain      32561
 capital-loss      32561
 hours-per-week    32561
 native-country    31978
 income            32561
dtype: int64

### Removing rows with Null values

In [32]:
income_ds.dropna(inplace=True)

In [33]:
income_ds.shape

(30162, 15)

* Now,there are 30162 rows ,so 2399 rows with missing values have been removed.

In [34]:
income_ds.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30162 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   age              30162 non-null  int64 
 1    workclass       30162 non-null  object
 2    fnlwgt          30162 non-null  int64 
 3    education       30162 non-null  object
 4    education-num   30162 non-null  int64 
 5    marital-status  30162 non-null  object
 6    occupation      30162 non-null  object
 7    relationship    30162 non-null  object
 8    race            30162 non-null  object
 9    sex             30162 non-null  object
 10   capital-gain    30162 non-null  int64 
 11   capital-loss    30162 non-null  int64 
 12   hours-per-week  30162 non-null  int64 
 13   native-country  30162 non-null  object
 14   income          30162 non-null  object
dtypes: int64(6), object(9)
memory usage: 2.6+ MB


In [35]:
income_ds.isnull().sum()

age                0
 workclass         0
 fnlwgt            0
 education         0
 education-num     0
 marital-status    0
 occupation        0
 relationship      0
 race              0
 sex               0
 capital-gain      0
 capital-loss      0
 hours-per-week    0
 native-country    0
 income            0
dtype: int64

* Now, There are not any null values in any of the attributes.

In [36]:
income_ds.describe()

age        fnlwgt   education-num   capital-gain  \
count  30162.000000  3.016200e+04    30162.000000   30162.000000   
mean      38.437902  1.897938e+05       10.121312    1092.007858   
std       13.134665  1.056530e+05        2.549995    7406.346497   
min       17.000000  1.376900e+04        1.000000       0.000000   
25%       28.000000  1.176272e+05        9.000000       0.000000   
50%       37.000000  1.784250e+05       10.000000       0.000000   
75%       47.000000  2.376285e+05       13.000000       0.000000   
max       90.000000  1.484705e+06       16.000000   99999.000000   

        capital-loss   hours-per-week  
count   30162.000000     30162.000000  
mean       88.372489        40.931238  
std       404.298370        11.979984  
min         0.000000         1.000000  
25%         0.000000        40.000000  
50%         0.000000        40.000000  
75%         0.000000        45.000000  
max      4356.000000        99.000000

## Data Formatting

In [37]:
income_ds[["age"," education-num"," capital-gain"," capital-loss"," hours-per-week"," fnlwgt"]]=income_ds[["age"," education-num"," capital-gain"," capital-loss"," hours-per-week"," fnlwgt"]].apply(pd.to_numeric)

In [38]:
income_ds.head()

age          workclass   fnlwgt   education   education-num  \
0   39          State-gov    77516   Bachelors              13   
1   50   Self-emp-not-inc    83311   Bachelors              13   
2   38            Private   215646     HS-grad               9   
3   53            Private   234721        11th               7   
4   28            Private   338409   Bachelors              13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

    capital-gain   capital-loss   hours-per-week  native-country  income  
0           2174              0               40   United-States   <=50K  
1              0              0               13   United-States   <=50K  
2              0              0               40   United-States   <=50K  
3              0              0               40   United-States   <=50K  
4              0              0               40            Cuba   <=50K